Fix 3084: `[Mobile]: Training alert appears before gesture completes`

codebuild-ro opened the pull request:
Fixes #3084

- Update logic to show command alert - now it is showing after gesture is finished and disappears after 5 seconds
- Added puppeteer testing
  - Verifies no alerts during gesture progress
  - Verifies alerts appear after gesture completion
  - Includes visual regression protection
  
### Cause
Actually, it was implemented for that behavior purposely, so it is difficult to consider it as a bug.
In one word, it was implemented for wrong behavior logic.
So I removed original code for that behavior and added similar code to other place which is for gesture ending handler.

raineorshine commented on src/e2e/puppeteer/__tests__/gesture-alert.ts:
```diff
@@ -0,0 +1,139 @@
+ * @param x2 - Ending X coordinate.
+ * @param y2 - Ending Y coordinate.
+ */
+const move = async (x1: number, y1: number, x2: number, y2: number): Promise<void> => {
```
This is a great idea, but you can use the existing [swipe](https://github.com/cybersemics/em/blob/9889f4a15288d82c787e2ea5b4e3ef976f8fe4f3/src/e2e/puppeteer/helpers/swipe.ts) helper. You'll just need to extend it to work in multiple directions.

raineorshine commented on src/e2e/puppeteer/__tests__/gesture-alert.ts:
```diff
@@ -0,0 +1,139 @@
+ *
+ * Note: No touchEnd() is called, so the gesture remains incomplete.
+ */
+const performIncompleteGesture = async () => {
```
I would prefer an API like `swipe('rdr')`

raineorshine commented on src/e2e/puppeteer/__tests__/gesture-alert.ts:
```diff
@@ -0,0 +1,139 @@
+    // Perform an incomplete gesture (no touchEnd)
+    await performIncompleteGesture()
+
+    // Wait a bit to ensure any potential alerts would have time to appear
+    await sleep(1000)
```
We do not allow arbitrary `sleep`, as it slows down the test suite for everyone on the team. You can wait for specific conditions, or `waitForFrames` if just waiting for rendering.

raineorshine commented on src/e2e/puppeteer/__tests__/__image_snapshots__/gesture-alert/alert-after-gesture-completion.png:
No snapshot tests for this. This is a behavioral bug, not a visual regression.

raineorshine reviewed (changes requested):
Hi, thanks for your submission.

What was the cause of the bug? That's the most important thing when it comes to bug fixes. Please clearly describe the cause of the bug. That will help me evaluate your proposed solution. Thanks!

codebuild-ro commented:
> Hi, thanks for your submission.
> 
> What was the cause of the bug? That's the most important thing when it comes to bug fixes. Please clearly describe the cause of the bug. That will help me evaluate your proposed solution. Thanks!

Actually, it was implemented for that behavior purposely, so it is difficult to consider it as a bug.
In one word, it was implemented for wrong behavior logic.
So I removed original code for that behavior and added similar code to other place which is for gesture ending handler.

raineorshine commented:
It's still possible to describe how the previous implementation failed to result in the correct behavior. I don't think it was ever intentional.

JoeDuong-dev commented:
@codebuild-ro Can you clarify the behavior this fix is addressing?

Update logic to show command alert - now it is showing after gesture is finished and disappears after 5 seconds

I tested on both mobile devices and the web (via DevTools), but the gesture panel does not disappear after 5 seconds as expected. For example, after swiping →↓ and holding, the panel appeared. I kept holding for longer than 5 seconds, and it remained visible.

With this branch, I do notice an improvement, the gesture panel triggers less often, but I still see the issue occasionally. This might be because I’m holding the ↓ gesture slightly too long. How long does the gesture need to be held to trigger the panel?

codebuild-ro commented:
Hi @JoeDuong-dev 
In this ticket, we need to show the alert after gesture finishes.
Seems you confused the alert as gesture panel.
<img width="400" height="726" alt="image" src="https://github.com/user-attachments/assets/2d2708b2-6986-484f-a693-72b574239b92" />

JoeDuong-dev commented:
Re-tested and confirmed that this is working as intended. I tested with various gesture combinations, and the alert displayed after each gesture. The panel then dismissed automatically after 5 seconds.

Everything is functioning according to the expected behavior.

Thanks @codebuild-ro for clarifying the confusion.
